guppylang/definition/pytket_circuits.py

```diff
@@ -0,0 +1,262 @@
+            self.input_circuit,
+            hugr_func,
+        )
+
+    def check_call(
+        self, args: list[ast.expr], ty: Type, node: AstNode, ctx: Context
+    ) -> tuple[ast.expr, Subst]:
+        """Checks the return type of a function call against a given type."""
+        # Use default implementation from the expression checker
+        args, subst, inst = check_call(self.ty, args, ty, node, ctx)
+        node = with_loc(node, GlobalCall(def_id=self.id, args=args, type_args=inst))
+        return node, subst
+
+    def synthesize_call(
+        self, args: list[ast.expr], node: AstNode, ctx: Context
+    ) -> tuple[ast.expr, Type]:
+        """Synthesizes the return type of a function call."""
+        # Use default implementation from the expression checker
+        args, ty, inst = synthesize_call(self.ty, args, node, ctx)
+        node = with_loc(node, GlobalCall(def_id=self.id, args=args, type_args=inst))
+        return node, ty
+
+
+@dataclass(frozen=True)
+class CompiledPytketDef(ParsedPytketDef, CompiledCallableDef):
+    """A function definition with a corresponding Hugr node.
+
+    Args:
+        id: The unique definition identifier.
+        name: The name of the function.
+        defined_at: The AST node where the function was defined.
+        ty: The type of the function.
+        python_scope: The Python scope where the function was defined.
+        input_circuit: The user-provided pytket circuit.
+        func_df: The Hugr function definition.
+    """
+
+    func_def: Node
+
+    def load_with_args(
+        self,
+        type_args: Inst,
+        dfg: DFContainer,
+        globals: CompiledGlobals,
+        node: AstNode,
+    ) -> Wire:
+        """Loads the function as a value into a local Hugr dataflow graph."""
+        # Use implementation from function definition.
+        return load_with_args(type_args, dfg, self.ty, self.func_def)
+
+    def compile_call(
+        self,
+        args: list[Wire],
+        type_args: Inst,
+        dfg: DFContainer,
+        globals: CompiledGlobals,
+        node: AstNode,
+    ) -> CallReturnWires:
+        """Compiles a call to the function."""
+        func_ty: ht.FunctionType = self.ty.instantiate(type_args).to_hugr()
+        type_args: list[ht.TypeArg] = [arg.to_hugr() for arg in type_args]
+        num_returns = len(type_to_row(self.ty.output))
+        call = dfg.builder.call(
+            self.func_def, *args, instantiation=func_ty, type_args=type_args
+        )
+        # Negative index slicing doesn't work when num_returns is 0.
+        if num_returns == 0:
+            return CallReturnWires(
+                regular_returns=[],
+                inout_returns=list(call[0:]),
+            )
+        else:
+            # Circuit function returns are the other way round to Guppy returns.
+            return CallReturnWires(
+                regular_returns=list(call[-num_returns:]),
+                inout_returns=list(call[:-num_returns]),
+            )
```

Comment:
Consider doing
```suggestion
        num_inouts = len(list(call)) - num_returns
        return CallReturnWires(
            regular_returns=list(call[inout_returns:]),
            inout_returns=list(call[:inout_returns]),
        )
```
to avoid the 0 edge case